solutions/USACO open20-haircut.mdx

```diff
@@ -0,0 +1,167 @@
+---
+id: usaco-1041
+source: USACO Gold 2020 US Open
+title: Haircut
+author: Jeffrey Meng
+---
+
+
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_haircut_gold_open20.html)
+
+First, we need to count the number of inversions. (Fun fact: the number of inversions is also a measure of how sorted an array is. If an array has 0 inversions, it's perfectly sorted in ascending order.) 
+
+One way to count inversions is to use a Binary Indexed Tree. Think of an inversion as a pair $i, j$, with $i < j$ & $A[i] > A[j]$. Then, we can easily compute, for each element in the array, the number of inversions it is $j$ for.
+
+To do this, we use a BIT as a frequency table, keeping track of the number of hairs before index $i$ of each possible value. If our BIT is named `tree`, `tree[i]` will tell us how many times $i$ is seen in the array of hairs. As long as we fill out the frequency table in the order the elements appear in the array, we can use the frequency table to also count the number of inversions, because the frequency table only contains hairs to the left of the hair we are currently considering.
+
+That leads us to this algorithm:
+
+1. Initialize a BIT so that the frequency of each value is 0.
+2. Create an array `inversions`. `inversions[j]` will store the number of inversion pairs with the second number being $j$
+3. For each hair $j$ in the order it appears in the array:
+
+    a. Count the number of taller hairs that appear before hair $j$ in the array (which is just the frequencies of all values greater than $A[j]$). Store this in `inversions[j]`.
```

Comment:
```suggestion
    a. Count the number of taller hairs that appear before hair $j$ in the array (which is just the frequencies of all values greater than $j$). Store this in `inversions[j]`.
```
j is the value, not the index

Reply:
j is the index, right?
we define the inversion as A[i] > A[j]

Reply:
> 3. For each hair $j$ in the order it appears in the array:
change this?
>     a. Count the number of taller hairs that appear before hair $j$ in the array (which is just the frequencies of all values greater than $j$). Store this in `inversions[j]`.
change inversions[j] to inversions[A[j]]

Reply:
Inversions tracks the id, but the BIT tracks the value.

Reply:
> inversionsWithValue[value] += numInversions;

i thought the inversions also uses the values as indexing

Reply:
whoops, you're right. The explanation was wrong. I updated it to make it more clear.